### PR TITLE
ER-324 Employer.Web Anonymous

### PR DESCRIPTION
**Pull request by chrisjensenuk**
Apologies for the huge PR. This ended up growing whilst we were wrestling with Locations and I ended up squashing the commits to make rebasing manageable.

- Created a MapImageHelper to centralise logic for creating links to Google Maps
- We don't show a pin (marker) in maps for anonymous vacancies. We just centre the map to the outcode
- Fixed PostcodeAnywhere lookup as it was not working correctly
- Renamed and extended `EmployerNameService` to `EmployerService`. This class is now responsible for returning the EmployerName and the EmployerDescription either from the vacancy or the employer profile.

**Geocoding Outcodes**
Extended the GeocodeService to add a new function to geocode outcodes. It tries these services in order to geocode the outcode portion of the postcode.
1. PoscodesIO
2. PostcodeAnywhere

I am not attempting to use existing vacancies to geocode due to the risk of potentially using a geocode that's accurate to the full postcode. As we are expecting only a small number of anonymous vacancies I thought it best to always use an external service to geocode the outcode.

### CODE REVIEW

File: src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs [hunk 1]
```diff
@@ -135,7 +135,37 @@ private void ValidateOrganisation()
                     .RunCondition(VacancyRuleSet.TradingName)
                     .WithRuleId(VacancyRuleSet.TradingName)
             );
-            
+
+            When(v => v.SourceOrigin == SourceOrigin.EmployerWeb && v.EmployerNameOption == EmployerNameOption.Anonymous, () =>
+                RuleFor(x => x.EmployerName)
+                    .NotEmpty()
+                    .WithMessage("You must provide a description")
```
Comment (dev4valtech) [inline src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs]: description?

File: src/Employer/Employer.Web/Extensions/EmployerNameOptionViewModelExtensions.cs [hunk 2]
```diff
@@ -20,6 +31,8 @@ public static EmployerNameOptionViewModel ConvertToModelOption(this EmployerName
                     return EmployerNameOptionViewModel.RegisteredName;
                 case EmployerNameOption.TradingName:
                     return EmployerNameOptionViewModel.ExistingTradingName;
+                case EmployerNameOption.Anonymous:
+                    return EmployerNameOptionViewModel.Anonymous;
                 default: 
                     throw new ArgumentOutOfRangeException();
```
Comment (sachpatel) [inline src/Employer/Employer.Web/Extensions/EmployerNameOptionViewModelExtensions.cs]: same here.

File: src/Employer/Employer.Web/ViewModels/Part1/EmployerName/EmployerNameOptionViewModel.cs [hunk 3]
```diff
@@ -6,6 +6,7 @@ public enum EmployerNameOptionViewModel
     {
         RegisteredName,
```
Comment (sachpatel) [inline src/Employer/Employer.Web/ViewModels/Part1/EmployerName/EmployerNameOptionViewModel.cs]: Can this be renamed to `EmployerIdentityOption` as it's slightly different to domain `EmployerNameOption` and also this isn't a ViewModel. Would suggest putting this in the Employer Web Models folder/namespace. `EmployerInfoModel` uses it so it is wrong that Models are referencing classes defined in the ViewModels namespace.

File: src/Employer/Employer.Web/Views/Part1/EmployerName/EmployerName.cshtml [hunk 4]
```diff
@@ -50,7 +49,41 @@
...
+                        <div class="govuk-radios__conditional" id="anonymous-conditional">
+                            <div class="govuk-form-group">
+                                <fieldset class="govuk-fieldset">
+                                    <div class="govuk-form-group">
+                                        <label class="govuk-label" asp-for="AnonymousName">
+                                            Brief description of what the employer does
+                                        </label>
+                                        <span class="govuk-hint">For example, car manufacturer or clothing retailer.</span>
+                                        <span esfa-validation-message-for="AnonymousName" class="govuk-error-message"></span>
+                                        <input asp-for="AnonymousName" data-val-length-max="200" class="govuk-input govuk-input--width-20" type="text">
```
Comment (sachpatel) [inline src/Employer/Employer.Web/Views/Part1/EmployerName/EmployerName.cshtml]: As there is a character limit and error message associated with max limit, would it be better for this field to be a text area as it is also asking for a description so a sentence.

Also there is no min limit meaning somebody could submit a single character such as `'`.

File: src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs [hunk 5]
```diff
@@ -135,7 +135,37 @@ private void ValidateOrganisation()
...
+                    .WithErrorCode("401")
+                    .MaximumLength(100)
+                    .WithMessage("The description must not exceed {MaxLength} characters")
+                    .WithErrorCode("402")
```
Comment (sachpatel) [inline src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs]: Are these error codes supposed to be the same as the ones below?

Also isn't the `description` in the error message quite generic in relation to the on screen field.

File: src/Shared/UnitTests/Vacancies.Client/Application/Services/EmployerServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    public class EmployerServiceTest
+    {
+        private Mock<IEmployerProfileRepository> _mockEmployerProfileRepository = new Mock<IEmployerProfileRepository>();
+
+        [Theory]
+        [InlineData(VacancyStatus.Live)]
+        [InlineData(VacancyStatus.Submitted)]
+        [InlineData(VacancyStatus.Approved)]
+        [InlineData(VacancyStatus.Closed)]
+        public async Task GetEmployerNameAsync_ShouldReturnVacancyEmployerName(VacancyStatus status)
```
Comment (dev4valtech) [inline src/Shared/UnitTests/Vacancies.Client/Application/Services/EmployerServiceTest.cs]: Can we break this into two classes, one for each method

File: src/Employer/Employer.Web/Orchestrators/VacancyPreviewOrchestrator.cs [hunk 7]
```diff
@@ -70,12 +70,9 @@ public async Task<OrchestratorResponse<SubmitVacancyResponse>> SubmitVacancyAsyn
...
             if (!vacancy.CanSubmit)
                 throw new InvalidStateException(string.Format(ErrMsg.VacancyNotAvailableForEditing, vacancy.Title));
-            
-            var employerProfile = await _vacancyClient.GetEmployerProfileAsync(vacancy.EmployerAccountId, vacancy.LegalEntityId);
 
-            // Update the vacancy with the current employer description text from Profile.
-            vacancy.EmployerDescription = employerProfile?.AboutOrganisation ?? string.Empty;
-            vacancy.EmployerName = await _vacancyClient.GetEmployerName(vacancy.Id);
+            vacancy.EmployerDescription = await _vacancyClient.GetEmployerDescriptionAsync(vacancy);
+            vacancy.EmployerName = await _vacancyClient.GetEmployerName(vacancy);
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Orchestrators/VacancyPreviewOrchestrator.cs]: 👍

File: src/Shared/Recruit.Shared.Web/Services/IGeocodeImageService.cs [hunk 8]
```diff
@@ -2,7 +2,7 @@
 {
     public interface IGeocodeImageService
     {
-        string GetMapImageUrl(string postcode, int imageWidth, int imageHeight);
-        string GetMapImageUrl(string latitude, string longitude, int imageWidth, int imageHeight);
+        string GetMapImageUrl(string postcode, int imageWidth, int imageHeight, bool showMarker);
+        string GetMapImageUrl(string latitude, string longitude, int imageWidth, int imageHeight, bool showMarker);
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Shared/Recruit.Shared.Web/Services/IGeocodeImageService.cs]: I think how it now is more explicit.  `canShowMarker` to me reads as "You can show a marker if you want to but its not necessary". `showMarker` is more of an explicit instruction to "show the marker"

[the remark this answers is not in the document]
Comment (sachpatel) [reply src/Shared/Recruit.Shared.Web/Services/IGeocodeImageService.cs]: 👍

File: src/Shared/UnitTests/Vacancies.Client/Application/VacancyValidation/SingleField/EmployerNameTests.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+using Esfa.Recruit.Vacancies.Client.Domain.Entities;
+using FluentAssertions;
+using Xunit;
+
+namespace Esfa.Recruit.UnitTests.Vacancies.Client.Application.VacancyValidation.SingleField
+{
+    public class EmployerNameTests : VacancyValidationTestsBase
+    {
+        [Fact]
+        public void ProviderWeb_RegisteredName_ShouldValidateEmpty()
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Shared/UnitTests/Vacancies.Client/Application/VacancyValidation/SingleField/EmployerNameTests.cs]: 👍 Looks like a hangover from when we only validated provider vacancies. Will change.

File: src/Employer/Employer.Web/ViewModels/ValidationMessages.cs [hunk 10]
```diff
@@ -9,7 +9,7 @@ public static class LevyDeclarationConfirmationMessages
 
         public static class LocationPreferenceMessages
         {
-            public const string SelectionRequired = "You must select one option.";
+            public const string SelectionRequired = "You must select a work address";
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/ViewModels/ValidationMessages.cs]: I know what you mean but its in the story.

File: src/Employer/Employer.Web/Views/Part1/EmployerName/EmployerName.cshtml [hunk 11]
```diff
@@ -50,7 +49,41 @@
...
                                 </div>
-                                <input asp-for="NewTradingName" class="govuk-input govuk-input--width-20" id="NewTradingName" name="NewTradingName" type="text" >
+                                <input asp-for="NewTradingName" class="govuk-input govuk-input--width-20" id="NewTradingName" name="NewTradingName" type="text">
+                            </div>
+                        </div>
+                        <div class="govuk-radios__divider">or</div>
+                        <div class="govuk-radios__item">
+                            <input asp-for="SelectedEmployerNameOption" value="Anonymous" data-aria-controls="anonymous-conditional" id="anonymous" class="govuk-radios__input" type="radio" >
+                            <label class="govuk-label govuk-radios__label" for="anonymous">
+                                do not display a name (be anonymous)
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Views/Part1/EmployerName/EmployerName.cshtml]: We're not going to show an employer name. We'll show the anonymous name which is more of a description.

File: src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs (hunk 1 above)
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs]: Its in the story. Employer name is a description of what the employer does not an alias.

File: src/Shared/Recruit.Vacancies.Client/Domain/Entities/Vacancy.cs [hunk 12]
```diff
@@ -116,5 +117,17 @@ public class Vacancy
...
+
+        /// <summary>
+        /// Is the employer anonymous
+        /// </summary>
+        public bool IsAnonymous => EmployerNameOption == Entities.EmployerNameOption.Anonymous;
+
+        /// <summary>
+        /// Should the vacancy be geocoded using the outcode part of the postcode
+        /// </summary>
+        public bool GeocodeUsingOutcode => IsAnonymous;
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Domain/Entities/Vacancy.cs]: `GeocodeVacancyCommandHandler` & `MapImageHelper` both need to be aware of how to geocode the vacancy so this decision logic needs to be shared and the vacancy is the common piece between these classes.

File: src/Shared/UnitTests/Vacancies.Client/Application/Services/EmployerServiceTest.cs [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+            result.Should().Be(employerName);
+        }
+
+        [Theory]
+        [InlineData(VacancyStatus.Draft)]
+        [InlineData(VacancyStatus.Referred)]
+        public async Task GetEmployerDescriptionAsync_ShouldReturnEmployerProfileAboutOrganisation(VacancyStatus status)
+        {
+            var employerDescription = "Employer Description";
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Shared/UnitTests/Vacancies.Client/Application/Services/EmployerServiceTest.cs]: 👍

File: src/Employer/Employer.Web/Extensions/EmployerNameOptionViewModelExtensions.cs [hunk 14]
```diff
@@ -8,8 +8,19 @@ public static class EmployerNameOptionViewModelExtensions
...
+            {
+                case EmployerNameOptionViewModel.RegisteredName:
+                    return EmployerNameOption.RegisteredName;
+                case EmployerNameOptionViewModel.ExistingTradingName:
+                    return EmployerNameOption.TradingName;
+                case EmployerNameOptionViewModel.NewTradingName:
+                    return EmployerNameOption.TradingName;
+                case EmployerNameOptionViewModel.Anonymous:
+                    return EmployerNameOption.Anonymous;
+                default:
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Extensions/EmployerNameOptionViewModelExtensions.cs]: 👍

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/VacancyClient.cs [hunk 15]
```diff
@@ -109,21 +109,14 @@ public Task<Vacancy> GetVacancyAsync(Guid vacancyId)
             return _repository.GetVacancyAsync(vacancyId);
         }
 
-        public Task<string> GetEmployerName(Guid vacancyId)
+        public Task<string> GetEmployerName(Vacancy vacancy)
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/VacancyClient.cs]: 👍

File: src/Employer/Employer.Web/Extensions/EmployerNameOptionViewModelExtensions.cs (hunk 2 above)
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Extensions/EmployerNameOptionViewModelExtensions.cs]: 👍

File: src/Employer/Employer.Web/ViewModels/Part1/EmployerName/EmployerNameOptionViewModel.cs (hunk 3 above)
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/ViewModels/Part1/EmployerName/EmployerNameOptionViewModel.cs]: 👍

File: src/Employer/Employer.Web/Views/Part1/EmployerName/EmployerName.cshtml (hunk 4 above)
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Views/Part1/EmployerName/EmployerName.cshtml]: This is as per the story. I'll chat to Jithesh about minlength.

File: src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs (hunk 5 above)
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Application/Validation/Fluent/FluentVacancyValidator.cs]: 👍 for the error codes.
As for `description`. Maybe, but its the content required by the story though.

File: src/Shared/UnitTests/Vacancies.Client/Application/Services/EmployerServiceTest.cs (hunk 6 above)
Comment (chrisjensenuk) [reply src/Shared/UnitTests/Vacancies.Client/Application/Services/EmployerServiceTest.cs]: 👍

**Pull request merged**